File: packages/zeroad-adapter/tools/camera-injector.cpp
// Camera Memory Injector for 0 A.D.
// Directly modifies camera position in memory during gameplay
// Usage: camera-injector.exe --pid <process_id> --x <x_coord> --z <z_coord> [--zoom <distance>]

#include <windows.h>
#include <iostream>
#include <string>
#include <cstring>
#include <cstdlib>

// Known 0 A.D. camera struct patterns (may need updating for different versions)
// Camera is typically stored as a C++ class with float members
struct CameraData {
    float x;
    float y;
    float z;
    float zoom;
    // ... other fields
};

// Global variables for parsed arguments
DWORD target_pid = 0;
float target_x = 0.0f;
float target_z = 0.0f;
float target_zoom = -1.0f;  // -1 = don't change
bool verbose = false;

void PrintUsage() {
    std::cout << "Camera Memory Injector for 0 A.D.\n";
    std::cout << "Usage: camera-injector.exe --pid <pid> --x <x> --z <z> [--zoom <zoom>] [--verbose]\n";
    std::cout << "Example: camera-injector.exe --pid 5432 --x 500 --z 800 --zoom 150\n";
}

bool ParseArguments(int argc, char* argv[]) {
    for (int i = 1; i < argc; i++) {
        std::string arg = argv[i];

        if (arg == "--pid" && i + 1 < argc) {
            target_pid = std::stoul(argv[++i]);
        }
        else if (arg == "--x" && i + 1 < argc) {
            target_x = std::stof(argv[++i]);
        }
        else if (arg == "--z" && i + 1 < argc) {
            target_z = std::stof(argv[++i]);
        }
        else if (arg == "--zoom" && i + 1 < argc) {
            target_zoom = std::stof(argv[++i]);
        }
        else if (arg == "--verbose") {
            verbose = true;
        }
        else if (arg == "--help" || arg == "-h") {
            PrintUsage();
            return false;
        }
    }

    if (target_pid == 0) {
        std::cerr << "Error: --pid is required\n";
        PrintUsage();
        return false;
    }

    return true;
}

// Scan memory for camera position signature
// This is a pattern scan approach similar to CheatEngine
DWORD_PTR FindCameraAddress(HANDLE process_handle) {
    // This is a simplified approach - in practice you'd need to:
    // 1. Scan for known signatures in pyrogenesis.exe
    // 2. Or use ASLR-safe relative offsets
    // 3. Or hook into known function pointers

    std::cout << "[Camera Injector] Scanning for camera data structure...\n";

    // NOTE: Real implementation would:
    // 1. Enumerate loaded modules
    // 2. Scan executable sections for patterns
    // 3. Use symbol information if available
    // 4. Try multiple offset strategies for different 0 A.D. versions

    // For now, return 0 and show error message
    std::cerr << "[ERROR] Camera address scanning not implemented.\n";
    std::cerr << "[INFO] To find camera address:\n";
    std::cerr << "  1. Use CheatEngine to find camera X coordinate\n";
    std::cerr << "  2. Note the address offset\n";
    std::cerr << "  3. Hardcode or pass as argument to this injector\n";

    return 0;
}

// Alternative: Use known offset (if camera structure is consistent)
bool InjectCameraPositionByOffset(HANDLE process_handle, DWORD_PTR base_address, float x, float z, float zoom = -1.0f) {
    if (base_address == 0) {
        std::cerr << "Invalid base address\n";
        return false;
    }

    if (verbose) {
        std::cout << "[*] Injecting camera position\n";
        std::cout << "    Base address: 0x" << std::hex << base_address << std::dec << "\n";
        std::cout << "    X: " << x << "\n";
        std::cout << "    Z: " << z << "\n";
        if (zoom >= 0) std::cout << "    Zoom: " << zoom << "\n";
    }

    // Write X coordinate (offset 0)
    if (!WriteProcessMemory(process_handle, (LPVOID)(base_address + 0), &x, sizeof(float), NULL)) {
        std::cerr << "Failed to write X coordinate\n";
        return false;
    }

    // Skip Y coordinate (offset 4) - usually not needed for camera

    // Write Z coordinate (offset 8)
    if (!WriteProcessMemory(process_handle, (LPVOID)(base_address + 8), &z, sizeof(float), NULL)) {
        std::cerr << "Failed to write Z coordinate\n";
        return false;
    }

    // Write zoom if specified (offset 12)
    if (zoom >= 0) {
        if (!WriteProcessMemory(process_handle, (LPVOID)(base_address + 12), &zoom, sizeof(float), NULL)) {
            std::cerr << "Failed to write zoom value\n";
            return false;
        }
    }

    if (verbose) {
        std::cout << "[*] Camera position injected successfully\n";
    }

    return true;
}

int main(int argc, char* argv[]) {
    std::cout << "0 A.D. Camera Memory Injector v1.0\n";
    std::cout << "==================================\n\n";

    if (!ParseArguments(argc, argv)) {
        return 1;
    }

    // Open target process
    HANDLE process_handle = OpenProcess(
        PROCESS_VM_READ | PROCESS_VM_WRITE,
        FALSE,
        target_pid
    );

    if (process_handle == NULL) {
        std::cerr << "Error: Could not open process " << target_pid << "\n";
        std::cerr << "  Make sure:\n";
        std::cerr << "  1. Process ID is correct\n";
        std::cerr << "  2. You have administrator privileges\n";
        std::cerr << "  3. The process is still running\n";
        return 1;
    }

    if (verbose) {
        std::cout << "[*] Successfully opened process " << target_pid << "\n";
    }

    // Find camera address in memory
    DWORD_PTR camera_address = FindCameraAddress(process_handle);

    if (camera_address == 0) {
        std::cerr << "\nCannot find camera address in process memory.\n";
        std::cerr << "\nOptions:\n";
        std::cerr << "1. Use CheatEngine to find the address:\n";
        std::cerr << "   - Run 0 A.D.\n";
        std::cerr << "   - Open CheatEngine\n";
        std::cerr << "   - Attach to pyrogenesis.exe\n";
        std::cerr << "   - Scan for current camera X position\n";
        std::cerr << "   - Note the address offset\n";
        std::cerr << "   - Update the offset in this tool\n\n";

        std::cerr << "2. Or provide offset as command-line argument (future enhancement)\n";

        CloseHandle(process_handle);
        return 1;
    }

    // Inject camera position
    bool success = InjectCameraPositionByOffset(process_handle, camera_address, target_x, target_z, target_zoom);

    CloseHandle(process_handle);

    return success ? 0 : 1;
}

File: packages/zeroad-adapter/tools/camera-injector.py
#!/usr/bin/env python3
"""
Camera Memory Injector for 0 A.D.

Directly modifies camera position in pyrogenesis.exe memory.
Works like CheatEngine but automated.

Requirements:
    pip install pymem

Usage:
    python camera-injector.py --x 400 --z 800 [--zoom 150] [--pid <pid>] [--verbose]

Example:
    python camera-injector.py --x 500 --z 600 --verbose
"""

import sys
import argparse
import struct
from typing import Optional, Tuple

try:
    import pymem
    import pymem.process
except ImportError:
    print("Error: pymem not installed")
    print("Install with: pip install pymem")
    sys.exit(1)


class CameraInjector:
    """Injects camera position into 0 A.D. process memory"""

    def __init__(self, verbose: bool = False):
        self.verbose = verbose
        self.pm: Optional[pymem.Pymem] = None
        self.camera_address: Optional[int] = None

    def log(self, message: str):
        """Print log message"""
        if self.verbose or True:  # Always log for now
            print(f"[Camera] {message}")

    def find_process(self) -> Optional[int]:
        """Find pyrogenesis.exe process ID"""
        try:
            import psutil

            for proc in psutil.process_iter(['pid', 'name']):
                try:
                    if 'pyrogenesis' in proc.info['name'].lower():
                        pid = proc.info['pid']
                        self.log(f"Found pyrogenesis.exe (PID: {pid})")
                        return pid
                except (psutil.NoSuchProcess, psutil.AccessDenied):
                    pass

            self.log("ERROR: pyrogenesis.exe not found")
            return None
        except ImportError:
            self.log("psutil not installed, cannot auto-find process")
            self.log("Usage: python camera-injector.py --pid <pid> --x <x> --z <z>")
            return None

    def attach(self, pid: Optional[int] = None) -> bool:
        """Attach to pyrogenesis process"""
        try:
            if pid is None:
                pid = self.find_process()
                if pid is None:
                    return False

            self.pm = pymem.Pymem(int(pid))
            self.log(f"Attached to process {pid}")
            return True
        except Exception as e:
            self.log(f"ERROR: Failed to attach: {e}")
            return False

    def find_camera_address(self) -> bool:
        """
        Find camera address in memory.

        This is the tricky part - we need to scan for a known pattern.
        For now, returns False to indicate manual setup needed.

        In practice:
        1. Use CheatEngine to find the address
        2. Set CAMERA_ADDRESS constant below
        3. Or pass as command-line argument
        """
        if self.pm is None:
            self.log("Process not attached")
            return False

        # Known offsets (these need to be found using CheatEngine)
        # Example: camera at base + 0xA5C8B0
        CAMERA_ADDRESS = 0x0  # Set this using CheatEngine results

        if CAMERA_ADDRESS == 0:
            self.log("ERROR: Camera address not configured")
            self.log("")
            self.log("To find camera address:")
            self.log("1. Open CheatEngine")
            self.log("2. Attach to pyrogenesis.exe")
            self.log("3. Scan for current camera X coordinate (as Float)")
            self.log("4. Filter by changing value in game")
            self.log("5. Note the address (e.g., 0x123ABC00)")
            self.log("6. Edit this file and set CAMERA_ADDRESS")
            self.log("")
            return False

        self.camera_address = CAMERA_ADDRESS
        self.log(f"Camera address: 0x{CAMERA_ADDRESS:X}")
        return True

    def set_position(self, x: float, z: float, zoom: Optional[float] = None) -> bool:
        """
        Write camera position to memory

        Camera struct layout (typical):
        - Offset +0: float X
        - Offset +4: float Y (usually skipped)
        - Offset +8: float Z
        - Offset +12: float zoom (optional)
        """
        if self.pm is None:
            self.log("Process not attached")
            return False

        if self.camera_address is None:
            self.log("Camera address not found")
            return False

        try:
            # Write X coordinate
            self.pm.write_float(self.camera_address + 0, x)
            self.log(f"Wrote X: {x}")

            # Write Z coordinate
            self.pm.write_float(self.camera_address + 8, z)
            self.log(f"Wrote Z: {z}")

            # Write zoom if specified
            if zoom is not None and zoom > 0:
                self.pm.write_float(self.camera_address + 12, zoom)
                self.log(f"Wrote Zoom: {zoom}")

            self.log(f"✓ Camera position updated: ({x}, {z})")
            return True
        except Exception as e:
            self.log(f"ERROR: Failed to write position: {e}")
            return False

    def close(self):
        """Close process handle"""
        if self.pm:
            self.pm.close()


def main():
    parser = argparse.ArgumentParser(
        description="0 A.D. Camera Memory Injector"
    )
    parser.add_argument('--x', type=float, required=True, help='Camera X coordinate')
    parser.add_argument('--z', type=float, required=True, help='Camera Z coordinate')
    parser.add_argument('--zoom', type=float, help='Camera zoom distance (optional)')
    parser.add_argument('--pid', type=int, help='Process ID (auto-find if not specified)')
    parser.add_argument('--verbose', action='store_true', help='Verbose output')

    args = parser.parse_args()

    injector = CameraInjector(verbose=args.verbose)

    # Attach to process
    if not injector.attach(args.pid):
        sys.exit(1)

    # Find camera address
    if not injector.find_camera_address():
        sys.exit(1)

    # Inject position
    if not injector.set_position(args.x, args.z, args.zoom):
        sys.exit(1)

    injector.close()
    print("\n✓ Camera injection complete")


if __name__ == '__main__':
    main()
